Statement eval followup refactor

neenaoffline opened the pull request.

neenaoffline committed:
Extract errors into errors ns

The intention is to move to a state of code where error types are
represented with symbols instead of strings and that they always
accompany the error's human-readable string `:representation`. This
should happen in future commits.

This commit also:
- Add some basic tests for the * operator.
- Extract code-errors from the errors ns
- Move some errors mark/reset/get code into the errors ns

neenaoffline committed:
Remove outdated TODOs

neenaoffline committed:
Rename all scratch -> code

neenaoffline committed:
Move from :value -> :scalar

neenaoffline committed:
Fix code error ui integration

This commit adds some tests for error handling in the code evaluation
flow. It also fixes the broken ui integration with code errors.

prabhanshuguptagit commented on src/bean/errors.cljs:
```diff
@@ -1,8 +1,38 @@
 (ns bean.errors)
 
+(defn reset [value]
+  (dissoc value :error))
+
+(defn mark [value error]
+  (merge value error))
+
+(defn get-error [{:keys [error]}]
+  error)
```
Lovely.

prabhanshuguptagit replied:
Lovely.

prabhanshuguptagit commented on src/bean/functions.cljs:
```diff
@@ -1,8 +1,8 @@
 
 (defn bean-concat [params]
-  (let [concated-str (reduce (fn [c {:keys [value]}] (str c value))
+  (let [concated-str (reduce (fn [c {:keys [scalar]}] (str c scalar))
```
Scalar doesn't apply well here. It usually implies there is also a vector that might have a direction. Couple of suggestions
- `eval-result`
- `ast-result`

Ruled out
- `return-value`
- `val`
- `interpreted-result`

I like `eval-result` since we're calling the ast evaluation `eval-ast`.

neenaoffline replied:
There are a couple of results as a result of evaluation. At least `:scalar`, `:matrix` and `:error`. Using the term scalar here in the sense of Scalars, Vectors & Matrices.

neenaoffline committed:
Remove superfluous intermediate error state handling

Previously in the bean codebase...

This function was taking a flag to suppress error escalation on the
initial evaluation, likely to account for a previous bug that I'm not
bothering to dig up right now.

If that last sentence sounds like it is full of foreshadowing, you're
right. It is.
